utils.py

```diff
@@ -4,84 +4,31 @@
-    tmk = typesystem.String(title="Master Key", max_length=16, min_length=16)
-
-class http_errors_or_ok:
-    """
-    returns dict of {"field": "error_message"}
-    pin length should be 4 (perhaps six?)
-    pan length: 16 or 19
-    twk, tmk: hex, 16
-    """
-    def __init__(self, request: dict):
-        self.request = request
-
-    def validate_common(self, field, **kwargs):
-        errors = {}
-        length = kwargs.get("length")
-        if self.request.get(field) == None:
-            errors[field] = "This field should not be empty"
-            return errors
-
-        if not isinstance(length, list):
-            length = list([length])
-
-        if not len(self.request.get(field)) in length:
-            errors[field] = f"{field} length should be in range {length}"
-            return errors
+# custom schema field
+class HexNumber(typesystem.String):
+    # add new error-code and error message 
+    typesystem.String.errors['hex_number'] = 'Must be a hexadicimal number'
 
-    def handle_keys(self, field, **kwargs):
-        length = kwargs.get("length")
-        errors = {}
-        if self.request.get(field) == None:
-            errors[field] = "This field should not be empty"
-            return errors
-
-        if not isinstance(length, list):
-            length = list([length])
-
-        if not len(self.request.get(field)) in length:
-            errors[field] = f"{field} length should be in range {length}"
-            return errors
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+    
+    def validate(self, value, **kwargs):
         try:
-            bytes.fromhex(self.request.get(field))
-        except ValueError:
-            errors[field] = "This field should be all hex"
-            return errors
+            bytes.fromhex(value)
+        except Exception as e:
+            raise self.validation_error("hex_number")
+        return super().validate(value, **kwargs)
 
-    def errors(self)->list:
-        pin_error = self.validate_common("pin", length=4)
-        pan_error = self.validate_common("pan", length=[19, 16])
-        twk_error = self.handle_keys("twk", length=16)
-        tmk_error = self.handle_keys("tmk", length=16)
-        errors = []
-        for err in [pin_error, pan_error, twk_error, tmk_error]:
-            if err:
-                errors.append(err)
-        return errors
-    
-    def has_errors(self):
-        if self.errors():
-            return True
-        return False
 
-def wants_json(request):
-    if request.headers["content-type"] == "application/json":
-        return True
-    return False
+class RequestFields(typesystem.Schema):
+    pin = typesystem.String(title="PIN", max_length=4, min_length=4)
+    pan = typesystem.String(title="PAN", max_length=19, min_length=16)
+    twk = HexNumber(title="Working Key", max_length=16, min_length=16)
+    tmk = HexNumber(title="Master Key", max_length=16, min_length=16)
 
```

Comment:
PERFECT! I only checked `typesystem` long ago when it first got launched with starllete. Glad they added these features to it rn.

One funny story noebs was first started with python before i stuck with the limitations of the `typesystem` limitations drawbacks and moved to go.